src/validators/shapes.rs

```diff
@@ -17,7 +19,54 @@ pub fn validate(gtfs: &gtfs_structures::Gtfs) -> Vec<Issue> {
             Issue::new(Severity::Error, IssueType::InvalidCoordinates, id)
                 .object_type(gtfs_structures::ObjectType::Shape)
         });
-    missing_coord.chain(valid).collect()
+    let valid_shape_id = gtfs
+        .trips
+        .iter()
+        .map(|(_id, trip)| create_invalid_shape_id_issue(trip, gtfs))
+        .filter_map(|o| o);
+
+    let used_shape_id: HashSet<String> = gtfs
+        .trips
+        .iter()
+        .map(|(_trip_id, trip)| trip.shape_id.to_owned())
+        .filter_map(|o| o)
+        .collect();
+
+    let existing_shape_id: HashSet<String> = gtfs
+        .shapes
+        .iter()
+        .map(|(shape_id, _shape)| shape_id.to_owned())
+        .collect();
+
+    let unused_shape_id = existing_shape_id.difference(&used_shape_id).map(|id| {
+        Issue::new(Severity::Warning, IssueType::UnusedShapeId, id)
+            .object_type(gtfs_structures::ObjectType::Shape)
+    });
```

Comment:
I tried to extract that into a separate method, but could not understand what the return type should be. It's an iterator of Issues, but the compiler thinks it's more complicated.

Reply:
if you have an idea @antoine-de I would be interested :)

Reply:
hum, I don't think it would be easy to do this. You could return an `impl Iterator<Item=Issue>` (or a `Box<dyn Iterator<Item=Issue>>`) but this method would need `used_shape_id` and `existing_shape_id`, and those would be temporary value of your method no?

I think there are ways to bypass this, but I guess keeping this in the `validate` method is the easy way out :grin:

Reply:
I wanted to write something like:
```rust
 let unused_shape_id = unused_shape_id(gtfs);
```
with the function defined elsewhere:
```rust
fn unused_shape_id(gtfs: ...) -> ?? {
    let used_shape_id: HashSet<String> = gtfs
        .trips
        .iter()
        .map(|(_trip_id, trip)| trip.shape_id.to_owned())
        .filter_map(|o| o)
        .collect();

    let existing_shape_id: HashSet<String> = gtfs
        .shapes
        .iter()
        .map(|(shape_id, _shape)| shape_id.to_owned())
        .collect();

    existing_shape_id.difference(&used_shape_id).map(|id| {
        Issue::new(Severity::Warning, IssueType::UnusedShapeId, id)
            .object_type(gtfs_structures::ObjectType::Shape)
    });

}
```

Reply:
yes but with this, you return an iterator over a temporary structure (used_shape_id and existing_shape_id will be destroyed at the end of the function)

Reply:
han ok I get it.